File: src/handlers/TrainHandler.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2016-02-01 14:19:07
# @Author  : RaymondWong (549425036@qq.com)
# @Link    : github/Raymond-Wong

import os
import sys
import numpy as np
sys.path.append('..')

import config.Const as Const
import utils.CommonUtil as CommonUtil

# 开始做迭代计算
def train(targetMatrix, userMatrix, itemMatrix, lr):
  # 计算当前误差
  # errorMatrix = targetMatrix - np.dot(userMatrix, itemMatrix)
  # 根据误差调整物品矩阵
  # newItemMatrix = itemMatrix + lr * (np.dot(errorMatrix.transpose(), userMatrix).transpose() - Const.OVERFITTING * itemMatrix)
  # 根据误差调整用户矩阵
  # newUserMatrix = userMatrix + lr * (np.dot(errorMatrix, itemMatrix.transpose()) - Const.OVERFITTING * userMatrix)
  errorMatrix = targetMatrix - np.dot(userMatrix, itemMatrix)
  for user in xrange(0, targetMatrix.shape[0]):
    for item in xrange(0, targetMatrix.shape[1]):
      if targetMatrix[user][item] == 0.0:
        continue
      for k in xrange(0, Const.FEATURES_AMOUNT):
        userMatrix[user][k] += lr * (errorMatrix[user][item] - Const.OVERFITTING * userMatrix[user][k])
        itemMatrix[k][item] += lr * (errorMatrix[user][item] - Const.OVERFITTING * itemMatrix[k][item])
  return (userMatrix, itemMatrix)

# 获取误差值
def getError(targetMatrix, userMatrix, itemMatrix):
  # tmpMatrix = targetMatrix - np.dot(userMatrix, itemMatrix)
  # tmpMatrix.shape = targetMatrix.shape
  # itemTmpMatrix = sum(itemMatrix)
  # userTmpMatrix = sum(userMatrix.transpose()).transpose()
  # userTmpMatrix.shape = (userTmpMatrix.shape[0], 1)
  # tmpMatrix = tmpMatrix * tmpMatrix + Const.OVERFITTING * (itemTmpMatrix * itemTmpMatrix) + Const.OVERFITTING * (userTmpMatrix * userTmpMatrix)
  # error = 0.0
  # for user in xrange(0, targetMatrix.shape[0]):
  #   for item in xrange(0, targetMatrix.shape[1]):
  #     if targetMatrix[user][item] != -2.0:
  #       error += tmpMatrix[user][item]
  # return error
  errorMatrix = targetMatrix - np.dot(userMatrix, itemMatrix)
  error = 0.0
  for user in xrange(0, targetMatrix.shape[0]):
    for item in xrange(0, targetMatrix.shape[1]):
      if targetMatrix[user][item] != 0.0:
        error += pow(errorMatrix[user][item], 2)
  return error

# 迭代调整矩阵
def iterateTrain(targetMatrix, userMatrix, itemMatrix):
  startKey = CommonUtil.startProcess('TrainHandler', '迭代调整矩阵')
  learing_rate = Const.LEARNING_RATE
  looper = 1
  lastError = None
  while True:
    (userMatrix, itemMatrix) = train(targetMatrix, userMatrix, itemMatrix, learing_rate)
    learing_rate *= 0.9
    thisError = getError(targetMatrix, userMatrix, itemMatrix)
    CommonUtil.logger("TrainHanlder", "INFO", str(looper) + " 次迭代的误差为 " + str(thisError))
    looper += 1
    # if lastError is not None and thisError >= lastError:
    if looper > Const.ITERATE_TIME:
      break
    lastError = thisError
  CommonUtil.endProcess(startKey)
  return (userMatrix, itemMatrix)

File: src/main/java/Executor/TrainHandler.java
package Executor;

import Config.Const;

/**
 * Created by raymondwong on 16-3-2.
 */
public class TrainHandler {

    /**
     * 根据误差调整对userMatrix和itemMatrix做一次调整
     * @param userMatrix
     * @param itemMatrix
     * @param targetMatrix
     * @param userAmount
     * @param itemAmount
     */
    public static void train(Float[][] userMatrix, Float[][] itemMatrix, Float[][] targetMatrix, int userAmount, int itemAmount) {
        for (int user = 0; user < userAmount; user++) {
            for (int item = 0; item < itemAmount; item++) {
                if (targetMatrix[user][item] == null) continue;
                Float error = targetMatrix[user][item];
                for (int k = 0; k < Const.FEATURES_AMOUNT; k++) {
                    error -= userMatrix[user][k] * itemMatrix[k][item];
                }
                for (int k = 0; k < Const.FEATURES_AMOUNT; k++) {
                    userMatrix[user][k] += Const.LEARNING_RATE * (error * itemMatrix[k][item] - Const.OVERFITTING * userMatrix[user][k]);
                    itemMatrix[k][item] += Const.LEARNING_RATE * (error * userMatrix[user][k] - Const.OVERFITTING * itemMatrix[k][item]);
                }
            }
        }
    }

    /**
     * 获取当前的错误
     * @param userMatrix
     * @param itemMatrix
     * @param targetMatrix
     * @param userSize
     * @param itemSize
     * @return
     */
    public static Float getError(Float[][] userMatrix, Float[][] itemMatrix, Float[][] targetMatrix, int userSize, int itemSize) {
        Float error = 0.0f;
        for (int user = 0; user < userSize; user++) {
            for (int item = 0; item < itemSize; item++) {
                if (targetMatrix[user][item] == null) continue;
                Float err = 0.0f;
                for (int k = 0; k < Const.FEATURES_AMOUNT; k++) {
                    err += userMatrix[user][k] * itemMatrix[k][item];
                    error += Const.OVERFITTING * (float)(Math.pow(userMatrix[user][k], 2) + Math.pow(itemMatrix[k][item], 2));
                }
                error += (float)Math.pow(targetMatrix[user][item] - err, 2);
            }
        }
        return error;
    }
}
